### ISSUE TITLE
Why does the notebook *require* terminado?

### ISSUE THREAD

**Issue by jdemeyer** [jupyter/notebook#1683]
In SageMath, we have always installed the `notebook` using `distutils`, i.e. doing `python setup.py install`. We did this without depending on `terminado` and it worked fine (except for the message `Terminals not available (error was No module named terminado)` shown when starting the notebook).

We are now switching to installation using `pip` (https://trac.sagemath.org/ticket/20218) and suddenly `terminado` is a requirement and Jupyter refuses to run without it.

I think that `terminado` is a cool gadget but I don't think it should be required to run Jupyter. The fix is easy: just remove the corresponding line from `setup.py`.

CC @SylvainCorlay

**Comment by Carreau** [comment]
IIRC we had too many users complaining that after installing notebooks they did not have the terminal available, and it make things consistent with all the other dependencies that are not really required to run the notebook. Technically you do not need IPython, or nbconvert to run the notebook. 

This was done in PR https://github.com/ipython/ipython/pull/7632.

**Comment by takluyver** [comment]
Yeah, these dependencies are really what Debian packaging calls 'recommends'. We don't have a way to express that distinction in Python packaging, so we make them dependencies.

You can pass a `--no-deps` option to pip to ignore dependencies, and take care of installing them yourself. Of course, you're then responsible for handling any changed dependencies on each upgrade. Or you can install normally and then `pip uninstall` any packages you don't want.

**Comment by jdemeyer** [comment]
> You can pass a --no-deps option to pip to ignore dependencies

It doesn't really *ignore* dependencies, it just installs the package without dependencies. With `--no-deps`, the dependencies are still recorded, giving errors at run-time instead of build-time.

**Comment by jdemeyer** [comment]
A more general issue with roughly *all* Jupyter/IPython projects is that they behave significantly different depending on whether they are installed using `python setup.py install` or `pip install PKG`. See https://github.com/ipython/ipywidgets/issues/718 for a particularly bad instance of this.

This is really annoying for SageMath, since we are changing the installation of Python packages from `python setup.py install` to `pip install .` and we needed a lot of fixes to Jupyter packages to install correctly this way. Other Python packages did not have this problem.

**Comment by takluyver** [comment]
Sorry about that. Running `setup.py install` directly has not been recommended for some time, because you get either:

1. The default behaviour of setuptools, which is a major cause of headaches
2. The behaviour of plain distutils, which may not do everything you want.

We've mostly held out from requiring setuptools, so with our projects you'll mostly get 2. However, we generally now consider pip the correct way to install packages, so in a sense you're just running into the modern systems.

> With --no-deps, the dependencies are still recorded, giving errors at run-time instead of build-time.

Obviously if they are really hard dependencies, you're going to get errors at run time if they're not installed. But if you're getting entry point scripts which use pkg_resources and check for dependencies, I would *strongly* recommend upgrading to a newer version of pip. Modern versions no longer create those setuptools-style entry points.

**Comment by jtpio** [comment]
Closing as this is likely not relevant anymore for the notebook package itself.

The terminals and the dependency on `terminado` are now handled in a separate extension: https://github.com/jupyter-server/jupyter_server_terminals

So feel free to open a new issue there if it's still needed, thanks!
